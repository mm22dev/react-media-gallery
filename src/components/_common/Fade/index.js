import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Spring } from 'react-spring/renderprops'
import { emptySpringSettings, identityFn, cloneObj } from '../../../lib/helpers'

class Fade extends Component {
  constructor (props) {
    super(props)
    this.createSpringSettings = this.createSpringSettings.bind(this)
    this.state = { show: false }
  }

  createSpringSettings (ComponentProps) {
    const fadeDirection = ComponentProps.inProp ? 'in' : 'out'
    const isAlreadyAnimated = this.state.show === ComponentProps.inProp
    const customDisplay = ComponentProps.customDisplayOnFadeOutEnd !== null ? ComponentProps.customDisplayOnFadeOutEnd : 'block'
    let springSettings = cloneObj(emptySpringSettings)
    delete springSettings.onRest
    const fromTo = {
      in: {
        from: {
          display: '',
          opacity: 0
        },
        to: {
          display: '',
          opacity: 1
        }
      },
      out: {
        from: { opacity: 1 },
        to: {
          display: isAlreadyAnimated ? customDisplay : '',
          opacity: 0
        }
      }
    }
    const { from, to } = fromTo[fadeDirection]
    springSettings.from = from
    springSettings.to = to
    springSettings.config = { duration: ComponentProps.duration ? ComponentProps.duration : 300 }
    springSettings.immediate = ComponentProps.immediate ? true : isAlreadyAnimated
    if (!isAlreadyAnimated) {
      springSettings.onRest = () => {
        this.setState({ show: ComponentProps.inProp }, () => ComponentProps.onFadeComplete())
      }
    }

    return springSettings
  }

  render () {
    const springSettings = this.createSpringSettings(this.props)

    return (
      this.props.children !== null
        ? <Spring {...springSettings}>
          {
            springProps => (
              React.cloneElement(this.props.children, {
                style: {
                  ...this.props.children.props.style,
                  ...springProps
                }
              })
            )
          }
        </Spring>
        : null
    )
  }
}

Fade.propTypes = {
  inProp: PropTypes.bool,
  duration: PropTypes.number,
  immediate: PropTypes.bool,
  children: PropTypes.node,
  customDisplayOnFadeOutEnd: PropTypes.string,
  onFadeComplete: PropTypes.func
}

Fade.defaultProps = {
  inProp: false,
  duration: 400,
  immediate: false,
  children: null,
  customDisplayOnFadeOutEnd: null,
  onFadeComplete: identityFn
}

export default Fade
