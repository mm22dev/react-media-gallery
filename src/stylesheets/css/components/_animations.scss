@mixin keyframes($name) {
    @-webkit-keyframes #{$name} {
        @content; 
    }
    @-moz-keyframes #{$name} {
        @content;
    }
    @-o-keyframes #{$name} {
        @content;
    }
    @-ms-keyframes #{$name} {
        @content;
    }
    @keyframes #{$name} {
        @content;
    } 
}

@mixin animationName($name){
    -webkit-animation-name: $name;
    -moz-animation-name: $name;
    -o-animation-name: $name;
    -ms-animation-name: $name;
    animation-name: $name;
}

@mixin animationDuration($duration){
    -webkit-animation-duration: $duration;
    -moz-animation-duration: $duration;
    -o-animation-duration: $duration;
    -ms-animation-duration: $duration;
    animation-duration: $duration;
}

@mixin animationDirection($direction){
    -webkit-animation-direction: $direction;
    -moz-animation-direction: $direction;
    -o-animation-direction: $direction;
    -ms-animation-direction: $direction;
    animation-direction: $direction;
}

@mixin animationIterationCount($iteration-count){
    -webkit-animation-iteration-count: $iteration-count;
    -moz-animation-iteration-count: $iteration-count;
    -o-animation-iteration-count: $iteration-count;
    -ms-animation-iteration-count: $iteration-count;
    animation-iteration-count: $iteration-count;
}