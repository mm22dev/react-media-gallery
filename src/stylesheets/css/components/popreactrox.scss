@import "./animations";

div#prt-container.prt-o-default {
    position: fixed;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    text-align: center;
    cursor: pointer;
    > div:nth-child(1) {
        display: inline-block;
        height: 100%;
        vertical-align: middle;
    }
    > div:nth-child(2) {
        position: absolute;
        left: 0;
        top: 0;
        width: 100%;
        height: 100%;
    }
}

div#prt-container > div.prt-p-default {
    display: none;
    vertical-align: middle;
    position: relative;
    z-index: 1;
    cursor: pointer;
}

div#prt-container > div.prt-p-default > div.prt-l-default {
    > div.animated {
        @include animationName(intermittence);
        @include animationDuration(.3s);
        @include animationDirection(alternate);
        @include animationIterationCount(infinite);
    }
}

@include keyframes(intermittence) {
    from { opacity: 0.05; }
    to { opacity: 0.5; } 
}