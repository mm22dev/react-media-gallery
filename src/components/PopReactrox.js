import React from 'react'
import PropTypes from 'prop-types'
import { PopupSettingsProvider } from '../providers/PopupSettings'
import Overlay from './ui/Overlay'
import '../stylesheets/css/components/popreactrox.scss'

const PopReactrox = ({ PRTSettings }) => (
  <PopupSettingsProvider PRTSettings={PRTSettings}>
    <Overlay />
  </PopupSettingsProvider>
)

PopReactrox.propTypes = {
  PRTSettings: PropTypes.object
}

PopReactrox.defaultProps = {
  PRTSettings: {}
}

export default PopReactrox
