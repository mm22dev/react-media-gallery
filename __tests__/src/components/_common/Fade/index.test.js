import { Fade } from '../../../../../src/components/_common'
import { compose } from 'redux'
import toJSON from 'enzyme-to-json'

const { shallow, mount } = Enzyme

describe('<Fade /> Component', () => {
  let wrapper
  let _click = jest.fn()
  let MockComponent = class MockComponent extends Component {
    constructor (props) {
      super(props)
      this.handleClick = this.handleClick.bind(this)
    }
    handleClick () {
      _click(this.props.style)
      return this.props.style
    }
    render () {
      return <div onClick={this.handleClick} className='mock-component'></div>
    }
  }

  it('Renders correctly', () => {
    compose(expect, toJSON, shallow)(<Fade inProp={true} immediate={true} customDisplayOnFadeOutEnd={'flex'} />).toMatchSnapshot()
  })

  describe('Assigning style to the inner Component', () => {
    let mountWrapper,
      fadeProps

    beforeAll(() => {
      fadeProps = {
        inProp: true,
        duration: 100,
        immediate: false
      }
      mountWrapper = () => {
        wrapper = mount(
          <Fade {...fadeProps }>
            <MockComponent />
          </Fade>
        )
      }
    })

    afterAll(() => jest.resetAllMocks())

    it('Works on animation start', () => {
      mountWrapper()
      wrapper.find(MockComponent).simulate('click')
      expect(_click).toHaveBeenCalledWith({ display: '', opacity: 0 })
    })

    it('Works on animation end', done => {
      fadeProps.onFadeComplete = () => {
        wrapper.find(MockComponent).simulate('click')
        expect(_click).toHaveBeenCalledWith({ display: '', opacity: 1 })
        done()
      }
      mountWrapper()
    })
  })

  describe('createSpringSettings - Core method of Component', () => {
    let mockedCompProps,
      _createSpringSettings

    beforeEach(() => {
      mockedCompProps = {
        inProp: false,
        duration: 200,
        immediate: false,
        customDisplayOnFadeOutEnd: null
      }
      _createSpringSettings = mount(<Fade />).instance().createSpringSettings
    })

    it('Assigns custom display rule on fadeOut', () => {
      mockedCompProps.customDisplayOnFadeOutEnd = 'flex'
      expect(_createSpringSettings(mockedCompProps))
        .toEqual({
          from: { opacity: 1 },
          to: {
            display: 'flex',
            opacity: 0
          },
          config: { duration: 200 },
          immediate: true
        })
    })

    it('Not re-executes animation if it has already been done', () => {
      expect(_createSpringSettings({ ...mockedCompProps }))
        .toEqual({
          from: { opacity: 1 },
          to: {
            display: 'block',
            opacity: 0
          },
          config: { duration: 200 },
          immediate: true
        })
    })

    it('Calls callback only on animation end', () => {
      let result

      result = _createSpringSettings({ ...mockedCompProps })
      expect(result.onRest).not.toBeDefined()

      mockedCompProps.inProp = true
      result = _createSpringSettings({ ...mockedCompProps })
      expect(result.onRest).toBeDefined()
    })
  })
})
