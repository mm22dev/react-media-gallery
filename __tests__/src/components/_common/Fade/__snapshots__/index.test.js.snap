// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fade /> Component Renders correctly 1`] = `""`;
